File: apps/web/src/lib/api/detection-export.ts
/**
 * Detection export API client for downloading CSV and ML datasets.
 *
 * spec/009 PR 4: both endpoints go through ``/web-api/v1`` (cookie +
 * CSRF session boundary). These are GET requests — no CSRF header is
 * required.
 */

import { apiClient } from './client';

const WEB_API_BASE = '/web-api/v1';

/**
 * Build query string from export filter params.
 */
function buildExportParams(params?: {
  status?: string;
  tag_id?: string;
  dataset_id?: string;
  detection_run_id?: string;
}): string {
  if (!params) return '';
  const query = new URLSearchParams();
  if (params.status) query.set('status', params.status);
  if (params.tag_id) query.set('tag_id', params.tag_id);
  if (params.dataset_id) query.set('dataset_id', params.dataset_id);
  if (params.detection_run_id) query.set('detection_run_id', params.detection_run_id);
  const str = query.toString();
  return str ? `?${str}` : '';
}

/**
 * Export detections as CSV file (triggers browser download).
 */
export async function exportDetectionsCSV(
  projectId: string,
  params?: { status?: string; tag_id?: string; dataset_id?: string; detection_run_id?: string }
): Promise<void> {
  const qs = buildExportParams(params);
  const response = await apiClient.requestRaw(
    `${WEB_API_BASE}/projects/${projectId}/detections/export/csv${qs}`
  );
  if (!response.ok) {
    throw new Error(`Export failed: ${response.statusText}`);
  }
  const blob = await response.blob();
  downloadBlob(blob, 'detections.csv');
}

/**
 * Export ML training dataset as ZIP file (triggers browser download).
 */
export async function exportMLDataset(
  projectId: string,
  params?: { dataset_id?: string; detection_run_id?: string }
): Promise<void> {
  const qs = buildExportParams(params);
  const response = await apiClient.requestRaw(
    `${WEB_API_BASE}/projects/${projectId}/detections/export/ml-dataset${qs}`
  );
  if (!response.ok) {
    throw new Error(`Export failed: ${response.statusText}`);
  }
  const blob = await response.blob();
  downloadBlob(blob, 'ml-dataset.zip');
}

/**
 * Trigger a browser file download from a Blob.
 */
function downloadBlob(blob: Blob, filename: string): void {
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = filename;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}

File: apps/api/echoroo/api/web_v1/projects/_detection_export.py
"""Project detection export BFF adapters (spec/009 PR 4).

Spec/009 PR 4 moves the two detection export endpoints from
``/api/v1`` to ``/web-api/v1``. The legacy ``/api/v1/detections.py``
handlers continue to own the streaming pipelines (CSV row-by-row +
mid-stream permission re-check for Hybrid Contract A-5, ZIP buffered
ML-dataset archive); the BFF layer is a thin adapter that lands the
request on the cookie + CSRF session boundary, fires
:func:`gate_action` once, and returns the legacy
:class:`fastapi.responses.StreamingResponse` unchanged.

Endpoints (2):

* GET ``/{pid}/detections/export/csv``        → ``DETECTION_EXPORT_CSV_ACTION``
* GET ``/{pid}/detections/export/ml-dataset`` → ``DETECTION_EXPORT_ML_DATASET_ACTION``

Both legacy handlers are already centrally gated through
:func:`gate_action`, so no entry is required in
``scripts/allowlists/permission_guard_allowlist.txt``. Both responses
return a binary stream (CSV / ZIP) and never name
``Recording`` / ``Detection`` / ``Site`` as a Pydantic response model,
so no ``response_filter`` allowlist entry is required either.
"""

from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Request
from fastapi.responses import StreamingResponse

from echoroo.api.v1 import detections as legacy_detections
from echoroo.core.actions import (
    DETECTION_EXPORT_CSV_ACTION,
    DETECTION_EXPORT_ML_DATASET_ACTION,
)
from echoroo.core.database import DbSession
from echoroo.core.permissions import gate_action
from echoroo.middleware.auth import CurrentUser
from echoroo.models.enums import DetectionStatus

router = APIRouter()


@router.get(
    "/{project_id}/detections/export/csv",
    summary="Export detections as CSV",
    description="BFF adapter for the legacy detection CSV export endpoint.",
)
async def export_detections_csv(
    project_id: UUID,
    request: Request,
    current_user: CurrentUser,
    db: DbSession,
    status: DetectionStatus | None = None,
    tag_id: UUID | None = None,
    dataset_id: UUID | None = None,
    detection_run_id: UUID | None = None,
) -> StreamingResponse:
    """Delegate detection CSV export to the legacy handler."""
    await gate_action(
        action=DETECTION_EXPORT_CSV_ACTION,
        project_id=project_id,
        current_user=current_user,
        request=request,
        db=db,
    )
    return await legacy_detections.export_csv(
        project_id=project_id,
        request=request,
        current_user=current_user,
        db=db,
        status=status,
        tag_id=tag_id,
        dataset_id=dataset_id,
        detection_run_id=detection_run_id,
    )


@router.get(
    "/{project_id}/detections/export/ml-dataset",
    summary="Export ML training dataset",
    description="BFF adapter for the legacy ML-dataset ZIP export endpoint.",
)
async def export_ml_dataset(
    project_id: UUID,
    request: Request,
    current_user: CurrentUser,
    db: DbSession,
    dataset_id: UUID | None = None,
    detection_run_id: UUID | None = None,
) -> StreamingResponse:
    """Delegate ML-dataset ZIP export to the legacy handler."""
    await gate_action(
        action=DETECTION_EXPORT_ML_DATASET_ACTION,
        project_id=project_id,
        current_user=current_user,
        request=request,
        db=db,
    )
    return await legacy_detections.export_ml_dataset(
        project_id=project_id,
        request=request,
        current_user=current_user,
        db=db,
        dataset_id=dataset_id,
        detection_run_id=detection_run_id,
    )
